framework/src/vectorpostprocessors/MaterialVectorPostprocessor.C

```diff
@@ -26,37 +26,47 @@ MaterialVectorPostprocessor::validParams()
                              "elements at the indicated execution points.");
   params.addRequiredParam<MaterialName>("material",
                                         "Material for which all properties will be recorded.");
-  params.addRequiredParam<std::vector<unsigned int>>(
-      "elem_ids", "Element IDs to print data for (others are ignored).");
+  params.addParam<std::vector<dof_id_type>>(
+      "elem_ids",
+      "Subset of element IDs to print data for. If omitted, all elements will be printed.");
   return params;
 }
 
 MaterialVectorPostprocessor::MaterialVectorPostprocessor(const InputParameters & parameters)
   : ElementVectorPostprocessor(parameters),
-    _elem_filter(getParam<std::vector<unsigned int>>("elem_ids").begin(),
-                 getParam<std::vector<unsigned int>>("elem_ids").end()),
     _elem_ids(declareVector("elem_id")),
-    _qp_ids(declareVector("qp_id"))
+    _qp_ids(declareVector("qp_id")),
+    _x_coords(declareVector("x")),
+    _y_coords(declareVector("y")),
+    _z_coords(declareVector("z"))
 {
   auto & mat = getMaterialByName(getParam<MaterialName>("material"), true);
   auto & prop_names = mat.getSuppliedItems();
   if (mat.isBoundaryMaterial())
     mooseError(name(), ": boundary materials (i.e. ", mat.name(), ") cannot be used");
 
-  for (auto & id : _elem_filter)
+  // Get list of elements from user
+  if (parameters.isParamValid("elem_ids"))
   {
-    auto el = _mesh.getMesh().query_elem_ptr(id);
-
-    // We'd better have found the requested element on *some*
-    // processor.
-    bool found_elem = (el != nullptr);
-    this->comm().max(found_elem);
-
-    // We might not have el on this processor in a distributed mesh,
-    // but it should be somewhere and it ought to have a material
-    // defined for its subdomain
-    if (!found_elem || (el && !mat.hasBlocks(el->subdomain_id())))
-      mooseError(name(), ": material ", mat.name(), " is not defined on element ", id);
+    const auto & ids = getParam<std::vector<dof_id_type>>("elem_ids");
+    _elem_filter = std::set(ids.begin(), ids.end());
```

Comment:
How is this valid C++?  I'm not saying "this isn't valid C++", because apparently it works with semi-recent g++ and clang++ with `-std=c++17`; I'm saying "I don't understand how this is valid C++".  `std::set<dof_id_type>(begin, end)` is a standard range constructor, but `std::set(begin, end)` I would have just sworn would be an undefined template parameter.  https://en.cppreference.com/w/cpp/language/template_argument_deduction has some new rules, sure, but I'm not seeing how they apply here.

(this is apropos of MOOSE compilation now failing on gcc 7.5, but IIRC we were planning to bump the MOOSE requirements to 8.something anyway)

Reply:
A great question...

Reply:
I've opened https://github.com/idaholab/moose/pull/25206

Reply:
(C++ novice says) I suppose I figured since the member variable was defined as `std::set<dof_id_type>`, further assignments could be deduced from the type.

Reply:
You might have got it right by accident?  By the time `set<dof_id_type>::operator=` comes into play, *some* set has been constructed from its right hand argument.  And that set has to be getting disambiguated based on `ids.begin()` and `ids.end()`.